crates/agent/src/controllers/collection.rs

```diff
@@ -0,0 +1,255 @@
+                    new_md5 = ?md5,
+                    "updating inferred schema"
+                );
+                let publication = publication_status
+                    .update_pending_draft("updating inferred schema".to_string(), control_plane);
+                let draft_row =
+                    publication
+                        .draft
+                        .collections
+                        .get_or_insert_with(&collection_name, || tables::DraftCollection {
+                            collection: collection_name.clone(),
+                            scope: tables::synthetic_scope(
+                                models::CatalogType::Collection,
+                                &collection_name,
+                            ),
+                            expect_pub_id: Some(state.last_pub_id),
+                            model: Some(collection_def.clone()),
+                        });
+                update_inferred_schema(draft_row, &schema)?;
+
+                let pub_result = publication_status
+                    .finish_pending_publication(state, control_plane)
+                    .await?;
+
+                if pub_result.publication_status.is_success() {
+                    self.schema_md5 = Some(md5);
+                    self.schema_last_updated = Some(pub_result.started_at);
+                } else {
+                    anyhow::bail!(
+                        "Failed to publish inferred schema: {:?}",
+                        pub_result.publication_status
+                    );
+                }
+            }
+        } else {
+            tracing::debug!(%collection_name, "No inferred schema available yet");
+        }
+
+        let next_run = Some(self.next_run(state.updated_at, control_plane.current_time()));
+        Ok(next_run)
+    }
+
+    fn next_run(&self, last_update: DateTime<Utc>, now: DateTime<Utc>) -> NextRun {
+        // The idea here is to check frequently if there isn't an inferred schema at all yet,
+        // so we can quickly start materializing some data. But after it works at least once,
+        // then we can slow down a little.
+        let min_backoff_minutes = if self.schema_md5.is_none() {
+            1i64
+        } else {
+            10i64
+        };
+
+        // We use a simple heuristic to determine how long to wait before
+        // checking the inferred schema again: how long has it been since
+        // the last time the inferred schema was updated. Then clamp that
+        // duration into a reasonable bounds. Collections that see very
+        // frequent inferred schema updates will be checked much more
+        // frequently, while those that are updated infrequently will be
+        // checked somewhat less frequently.
+        let start_time = self.schema_last_updated.unwrap_or(last_update);
+        let after_minutes = now
+            .signed_duration_since(start_time)
+            .num_minutes()
+            .max(min_backoff_minutes)
```

Comment:
I think this is actually right?
Starting with the duration `d` since the previous update of the inferred schema, if `d < min_backoff_minutes` then we want to use `min_backoff_minutes`. But if `d > 120 minutes`, then we take the minimum value of `120`.